### PR TITLE
Universal/point light shadows

### PR DESCRIPTION
**Pull request by Wilfrid-Unity**
### Purpose of this PR

This pull request adds support for real-time Point Light shadows in Universal RP, as well as some fixes for Universal RP real-time shadows.
It brings Unity's Universal RP closer to parity with the Built-In RP.

**Main changes**

Point Light shadows are implemented by extending pre-existing class _UnityEngine.Rendering.Universal.Internal.AdditionalLightsShadowCasterPass_, that was previously used only for Spot Lights shadows.

In Unity code the term "shadow slice" is used to refer to shadow map textures generated for a single shadow casting light.
While a Spot Lights only require 1 shadow slice, for Point Lights 6 shadow slices are needed (one for each face of a cube around the light). 
Therefore the main changes in this pull request are related to the indexing of those additional shadow slices.

**Other additions from this pull request**

- When applying Punctual Lights Shadow Normal Bias, the light direction used is taken separately for each different shadow caster vertex position. Previously, the same light direction was used for all vertices, which caused the shape of shadows to change when a point light was rotated.

- Punctual light shadow setting "Soft Shadow" or "Hard Shadow" can now be set independently for each light. Previously, assigning "Soft Shadows" to one punctual light in the scene forced all shadow-casting punctual lights in the scene to use the "Soft Shadows" setting.

- The number of shadow casters rendered when generating spot light shadow maps is reduced. Previously, culling settings caused excess casters to be rendered, with an impact on perfomance.

**Not addressed by this pull request**

- Per-light shadow resolution override is not implemented by this pull request. It will be added separately.

---
### Testing status

**Manual Tests**:
I manually verified that punctual light shadows work in several scenarios combining multiple types of lights and shadow settings.
I manually verified that point light shadows were correctly rendered when using the Deferred Renderer.
I manually verified that point light shadows were correctly rendered when using Structured Buffers instead of Uniform Buffers, to pass shadow information to the GPU during the Lighting pass.

**Automated Tests**: 
This pull request adds test "145_ManyAdditionalLightShadowTypes" to suite UniversalGraphicsTests:

This scene contains 4 punctual lights of various types and shadow settings. Its purpose is to detect regressions related to the indexing of punctual light shadows.
I manually checked that changing shadow type for lights in the scene, caused the test to detect a failure.

**Yamato**: 

Automated test results are available here (no regression introduced to Universal Graphics and VFX tests):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fpoint-light-shadows

I also manually checked that Universal Hybrid tests were passing.

**Performance Tests**:

Here are numbers describing the performance impact of enabling point light shadows in a benchmark scene with a high number of polygons:
- On mobile platforms: https://docs.google.com/spreadsheets/d/1BwQxOJXhzk4X6GeSvPqUlbuzKSFYtH_N_oIpxKsvtbg/edit#gid=861882497
- On mobile console: https://docs.google.com/spreadsheets/u/1/d/1cx2uzkn6fmlsZY64DQWu8LtjY__KoLnGDKYRugfLb6E

Overall, the cost of adding Universal RP Point Light Shadows to the scene, is less than the cost of adding Built-In RP Point Light Shadows to the same scene.

---

### Comments to reviewers

**Other links related to this effort**
- Entry for Point Light Shadows on Universal RP roadmap https://portal.productboard.com/8ufdwj59ehtmsvxenjumxo82/c/14-point-light-shadows
- User discussion about Universal RP Point Light Shadows on Unity forums: https://forum.unity.com/threads/urp-point-light-shadows.881902/

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs:172]: I would like to understand this a bit better. Are you sure the additionally culled casters are valid? And no artifacts are produced.

**Review by eh-unity — COMMENTED** [review]
Looks very good to me. I really liked that you generously added comments and clarified existing behavior.
All the things I commented are relatively minor issues.

File: com.unity.render-pipelines.universal/Runtime/ShadowUtils.cs
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Runtime/ShadowUtils.cs]: You're right, it's better to avoid changing the meaning of constant variables.
Thanks for the suggestions, I pushed change 7a92566 where _LightPosition and _LightDirection are separate variables named more appropriately.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs]: That makes sense, I pushed revision 7a92566 where "Directional" is the implicit default case, and "_CASTING_PUNCTUAL_LIGHT_SHADOW" the variation.

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree7CommonInput.hlsl
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree7CommonInput.hlsl]: Great suggestions, I pushed change 7a92566 where _LightPosition and _LightDirection are separate variables named more appropriately. On many platforms the unused constant should be optimized out by the shader compiler.

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:304]: Thanks for the recommendation, I did that in revision 36c3908

File: com.unity.render-pipelines.universal/ShaderLibrary/ShaderTypes.cs
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/ShaderTypes.cs]: Thanks for the recommendation, I did that in revision 36c3908

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:345]: Do you mean moving this section to a new ShadowDeprecated.hlsl file?

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs:172]: This comment summarizes the conclusion I came to after shortly experimenting with [ScriptableRenderContext.DrawShadows](https://docs.unity3d.com/2020.2/Documentation/ScriptReference/Rendering.ScriptableRenderContext.DrawShadows.html) parameters.
If you would like to investigate further this experiment, I uploaded the test projects I used, and added a bit of explanations, here:
https://github.com/Wilfrid-Unity/main-light-shadow-cascades-culling

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl]: I like the temp variable approach, I added that in 0b5583d

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs
Comment (ellioman) [reply com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs:172]: Should this be filed as a bug?

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs
Comment (ellioman) [inline com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:25]: We've recently started moving deprecated things into one of:
- Editor/Deprecated.cs
- Runtime/Deprecated.cs
- Deprecated.hlsl
Depending on where that code lives.

For the .cs files we do so by making the class partial, and creating another partial partition in the Deprecated file. This would make sense to move there as well. Then in 11.x we will change this to [Obsolete(""), true] and remove it in 12.x.

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs
Comment (ellioman) [inline com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:26]: We've recently started moving deprecated things into one of:
- Editor/Deprecated.cs
- Runtime/Deprecated.cs
- Deprecated.hlsl

Depending on where that code lives.

For the .cs files we do so by making the class partial, and creating another partial partition in the Deprecated file. This would make sense to move there as well. Then in 11.x we will change this to [Obsolete(""), true] and remove it in 12.x.

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl
Comment (ellioman) [inline com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:288]: This will be a breaking change for anyone currently calling `AdditionalLightRealtimeShadow` by adding the lightDirection.

**Review by ellioman — CHANGES_REQUESTED** [review]
Few comments.
There's a breaking API change in here.
I'm also wondering about what needs to be done on the ShaderGraph side.

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl]: I moved this deprecated variable declaration to the Deprecated.hlsl file in commit dabaf72

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs]: Thanks for the explanations.
I moved this deprecated variable declaration to the Deprecated.cs file in commit dabaf72

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:288]: Good catch.
To prevent user shader compilation errors, I added a version of the old API in commit dabaf72

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs:172]: I think so. I filed this issue [1273362](https://issuetracker.unity3d.com/product/unity/issues/guid/1273362) about it.

**Comment by Wilfrid-Unity** [comment]
> Few comments.
> There's a breaking API change in here.
> I'm also wondering about what needs to be done on the ShaderGraph side.

@ellioman Thanks for the comments.
I moved the deprecated declarations.

Regarding Shader Graph, here is the result of a simple manual test that I did.
The material of the sphere on the left uses a PBR/Lit shader graph shader.
The sphere on the right uses the Lit shader provided in the URP package.
Point light shadow seemed to be projected correctly over both materials:
image

Do you know who could be added as a Reviewer or QA regarding the impact of those changes on ShaderGraph?

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl
Comment (ellioman) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:288]: Great! You can move the old API to deprecated.hlsl :)

**Comment by ellioman** [comment]
> Do you know who could be added as a Reviewer or QA regarding the impact of those changes on ShaderGraph?

I've added ShaderGraph as a reviewer now to make sure things work 100% there as well.

File: com.unity.render-pipelines.universal/ShaderLibrary/ShaderTypes.cs
Comment (ellioman) [reply com.unity.render-pipelines.universal/ShaderLibrary/ShaderTypes.cs]: Ahh, this is a public struct in a public static class so this is a breaking change.
We need to move it to Deprecated.cs and mark it as obsolete.

File: com.unity.render-pipelines.universal/Shaders/ShadowCasterPass.hlsl
Comment (alindmanUnity) [inline com.unity.render-pipelines.universal/Shaders/ShadowCasterPass.hlsl:38]: please make sure that any changes to the passes here are also reflected in the ShaderGraph includes, in Editor/Shader Graph/Includes

File: com.unity.render-pipelines.universal/Shaders/SimpleLit.shader
Comment (alindmanUnity) [inline com.unity.render-pipelines.universal/Shaders/SimpleLit.shader:132]: do we need to add this pragma to shader graph generation?

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:288]: I tried to move the old `AdditionalLightRealtimeShadow(int lightIndex, float3 positionWS)` API to Deprecated.hlsl, but since it references the new version of the API, it causes compilation errors in shaders that include Deprecated.hlsl but not Shadows.hlsl.

Do you think Shadows.hlsl should be included into every shader that refers to Deprecated.hlsl? This might make them a bit slower to process at build time ; and on platforms that support it, could make low-level shader debugging a bit more cumbersome.

Comment (ellioman) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:288]: Ahh I see. No let's then keep it the way it is. 
I will take up the discussion of removing deprecated things after the next major bump. And since you commented it as deprecated I will find it once we get to that point :)
All good!

File: com.unity.render-pipelines.universal/ShaderLibrary/ShaderTypes.cs
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/ShaderTypes.cs]: OK, I moved the `ShaderInput.ShadowData` into Deprecated.cs with revision 5664126.
I also moved the hlsl-side declaration to Deprecated.hlsl.

At first I tried to move `ShaderInput.ShadowData` into com.unity.render-pipelines.universal/Runtime/Deprecated.cs but this caused type conflicts between assemblies because `ShaderInput.ShadowData` is in assembly `Unity.RenderPipelines.Universal.ShaderLibrary` while com.unity.render-pipelines.universal/Runtime/Deprecated.cs is in assembly `Unity.RenderPipelines.Universal.Runtime`.
Moving `ShaderInput.ShadowData` into a new com.unity.render-pipelines.universal/ShaderLibrary/Deprecated.cs file included in the same `Unity.RenderPipelines.Universal.ShaderLibrary` assembly, solved the issues.

File: com.unity.render-pipelines.universal/Shaders/ShadowCasterPass.hlsl
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Shaders/ShadowCasterPass.hlsl:38]: Thanks for the comment, I updated the ShaderGraph headers to include this change as well (revision 05ea77e).

I verified that replacing the materials used in test scene TestProjects/UniversalGraphicsTest/Assets/Scenes/145_ManyAdditionalLightShadowTypes.unity by materials using ShaderGraph shaders, produced the same visual results.

I also verified that the changes in this pull request branch did not change the rendering of the scene in https://github.com/Unity-Technologies/ShaderGraph-Custom-Lighting

File: com.unity.render-pipelines.universal/Shaders/SimpleLit.shader
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Shaders/SimpleLit.shader:132]: Yes, this pragma is needed in shadow-casting shader passes, in order to have Normal Bias applied correctly for punctual light shadows.

In revision 05ea77e I updated shader graph code that generates shadow caster pass shader, to make it include this keyword.

File: com.unity.render-pipelines.universal/CHANGELOG.md
Comment (ellioman) [inline com.unity.render-pipelines.universal/CHANGELOG.md:31]: `### Added` needs to be at the top of the 10.1 list. And there should be only one, there are three right now for some reason. You probably need to merge the latest from master.

File: com.unity.render-pipelines.universal/Runtime/ShadowUtils.cs
Comment (ellioman) [inline com.unity.render-pipelines.universal/Runtime/ShadowUtils.cs:40]: Do we maybe need to keep the old variation of that function in case users are calling it and want the view and proj matrices?

**Review by ellioman — APPROVED** [review]
Added some minor comments.
Also, we need to make sure we strip `_CASTING_PUNCTUAL_LIGHT_SHADOW` if it's not in use. The place to do that is inside `com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs`.

I'm approving this PR assuming that gets added. @simon-engelbrecht-soerensen Make sure to test that it gets stripped when @Wilfrid-Unity adds it :)

File: com.unity.render-pipelines.universal/CHANGELOG.md
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/CHANGELOG.md]: Thanks, I did that in 0a41462, and later updated versions in e716c26 after a conversation with @oleks-k

File: com.unity.render-pipelines.universal/Runtime/ShadowUtils.cs
Comment (Wilfrid-Unity) [reply com.unity.render-pipelines.universal/Runtime/ShadowUtils.cs:48]: Yes that might be happening, I added it back in 3c68448

**Comment by Wilfrid-Unity** [comment]
> Added some minor comments.
> Also, we need to make sure we strip `_CASTING_PUNCTUAL_LIGHT_SHADOW` if it's not in use. The place to do that is inside `com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs`.
> 
> I'm approving this PR assuming that gets added. @simon-engelbrecht-soerensen Make sure to test that it gets stripped when @Wilfrid-Unity adds it :)

Thanks @ellioman for the additional review, and for the help about shader stripping.

In 2184cd9 I added code that strips ShadowCaster shader variants if Main or Additional light shadow features are disabled in the Universal RP asset.
I verified in shader variant logs that the number of variants was indeed reduced in those cases.

File: .yamato/ruamel/jobs/abv/abv_all_project_ci_nightly.py
Comment (sophiaaar) [inline .yamato/ruamel/jobs/abv/abv_all_project_ci_nightly.py:23]: i suspect the changes to the yml and py files are the result of a bad merge - please resolve this before landing to master

**Review by sophiaaar — CHANGES_REQUESTED** [review]

File: .yamato/ruamel/jobs/abv/abv_all_project_ci_nightly.py
Comment (Wilfrid-Unity) [reply .yamato/ruamel/jobs/abv/abv_all_project_ci_nightly.py]: Thanks @sophiaaar that was a bad merge indeed.
I fixed the merge in revision cd8d544
Those .yml and .pm [file changes](https://github.com/Unity-Technologies/Graphics/pull/1616/files) do not appear anymore.

Comment (sophiaaar) [reply .yamato/ruamel/jobs/abv/abv_all_project_ci_nightly.py]: thank you! approved for this

**Review by sophiaaar — APPROVED** [review]

**Review by ryanhy-unity — APPROVED** [review]
Tested new test scene using Oculus on D3D11 (multipass and instancing), OpoenGLES3 and Vulkan (both Multiview) (android) on Rift S and Quest and the scene looked correct for every case.

**Review by alindmanUnity — APPROVED** [review]

**Comment by vaidasma** [comment]
Tested on devices:
N/A, Xiaomi Mi A2 (Mi A2), Android 9, CPU: Snapdragon 660, GPU: Adreno (TM) 512
VLNQA00267, Samsung Galaxy S10+ (SM-G975F), Android 10, CPU: NOT FOUND, GPU: Mali-G76
VLNQA00230, Oneplus OnePlus6T (ONEPLUS A6013), Android 9, CPU: Snapdragon 845 SDM845, GPU: Adreno (TM) 630
VLNQA00010, Samsung Galaxy S8 (SM-G950U), Android 7.0, CPU: Snapdragon 835 MSM8998, GPU: Adreno (TM) 540
VLNQA00326, Samsung Galaxy Note10+ (SM-N975W), Android 10, CPU: Snapdragon 855 SM8150, GPU: Adreno (TM) 640
VLNQA00011, Xiaomi Mi 5s (MI 5s), Android 6.0.1, CPU: Snapdragon 820 MSM8996, GPU: Adreno (TM) 530
VLNQA00237 iPhone 6S A7 13.7
VLNQA00190 iPhone 5S A9 12.4.8

Tested gfx API's: Gles3, Vulkan, Metal

Project used for testing (modified to include realtime point lights with settings' variations):
https://github.com/ernestasKupciunas/Winter
https://github.cds.internal.unity3d.com/aleksandras-dzikia/mobile_realtime_lighting

Any shadow settings did not seem to produce glitches or any other visual artifacts.
Exception being one device:

 Oneplus OnePlus6T - crashes when GLES3 is set as a gfx api (on Mono and IL2CPP). attaching a logcat

[StackTraceGles3OnePlus6T.txt](https://github.com/Unity-Technologies/Graphics/files/5386119/StackTraceGles3OnePlus6T.txt)

**Comment by Wilfrid-Unity** [comment]
Thanks a lot for the tests on mobiles @vaidasma 


> Oneplus OnePlus6T - crashes when GLES3 is set as a gfx api (on Mono and IL2CPP). attaching a logcat

Are you able to find exactly which of the [commits](https://github.com/Unity-Technologies/Graphics/pull/1616/commits) introduces the crash on OnePlus6T GLES3?

**Comment by Wilfrid-Unity** [comment]
> Oneplus OnePlus6T - crashes when GLES3 is set as a gfx api (on Mono and IL2CPP). attaching a logcat

Thank you very much @vaidasma for your help with remote testing on OnePlus device.
I found that the issue happened from revision 2184cd9, it looks like stripping GLES shaders causes the issue with this device,

I added work-around 5330ff9 and could run the repro project on OnePlus6T after that.

**Review by vaidasma — APPROVED** [review]
The OnePlus6T crashing issues have been resolved.

**Review by simon-engelbrecht-soerensen — APPROVED** [review]
Last remaining issues I found have been fixed.
Test doc [here](https://docs.google.com/document/d/1tkQdYKx2bED5hp3dTmUlkeZS_moiUsRZeMS4hMXHfXk)

**Review by eh-unity — APPROVED** [review]

**Review by Verasl — APPROVED** [review]
LGTM

**Comment by ellioman** [comment]
Yamato failures not caused by this PR. Good to merge.

**Pull request merged**
